### PR TITLE
[KEYCLOAK-16780] - Allow batching writes to storage when running migration

### PR DESCRIPTION
**Pull request by pedroigor**
These changes aim to:

* Allow setting whether migration changes should be done in batches
* Allow setting how many migration changes should be done in batch
* Possible useful to other scenarios where a large number of changes to storage is made

Expected result:

* Time to migrate should be drastically reduced. Tests using the latest version migrate 1K realms in `555528ms`, whereas with these changes migration happens in ` 54595ms`.

During migration, we should expect degradation due to the number of changes being made. By using a batch approach (as recommended by Hibernate, btw), we can improve performance and startup.

We should expect better results if running using https://github.com/keycloak/keycloak/pull/7712.

### CODE REVIEW

File: model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java:40]:
```suggestion
    private static final Pattern WRITE_METHOD_NAMES = Pattern.compile("persist|merge");
```

File: model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java:69]:
```suggestion
            if (WRITE_METHOD_NAMES.matcher(method.getName()).matches()) {
```

File: model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java:44]: prefer primitive types when the values cannot be `null` so that the JVM can optimize more easily

**Review by hmlnarik — COMMENTED** [review]
Thanks @pedroigor , looks good to me. This seems a good idea and I like the fact that this is limited to the migration. Have you tried any different limit, e.g. 500?

**Comment by hmlnarik** [comment]
Also, has the pipeline been run on this PR? Since it touches migration, I'd rather have its results prior to merging.

**Comment by mposolda** [comment]
@pedroigor Thanks!

Will be good to have pipeline results for migration tests as @hmlnarik mentioned.

Also I wonder that we may still reach the limits during migration for example with 5000 realms? I wonder that we may also need to increase the transaction timeout for the migration transaction? This can be done with KeycloakModelUtils.runJobInTransactionWithTimeout . But maybe can be done as a follow-up (if needed), so I don't have anything against merge this PR as is.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java
Comment (vramik) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java:52]: Do we have some performance results with different `STORAGE_BATCH_SIZE` values? I think we should set the default value based on best results, wdyt?

**Review by vramik — COMMENTED** [review]
Thank you @pedroigor for the PR. Great work, I have just one question.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java
Comment (pedroigor) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java:52]: Sure. That number is what worked well for the use case we are testing using 1k realms.

Personally, I don't think it is critical for this PR. We can always change the value and users are free to set any value they want.

There is more work in the backlog that could help with this. But they are a follow-up to this PR and the other one related to multi-realm.

**Comment by pedroigor** [comment]
> Also I wonder that we may still reach the limits during migration for example with 5000 realms? I wonder that we may also need to increase the transaction timeout for the migration transaction? This can be done with KeycloakModelUtils.runJobInTransactionWithTimeout . But maybe can be done as a follow-up (if needed), so I don't have anything against merge this PR as is.

It works very well for our current scenario using 1k realms. I won't be surprised if running 5k the results are also good.

This is already an improvement. So I think any other could be done as a follow up.

**Review by vramik — APPROVED** [review]

**Comment by vramik** [comment]
Thanks @pedroigor, let's wait for the pipeline results. If those are ok, changes looks good to me.

**Comment by mposolda** [comment]
@pedroigor +1 to do as a follow-up if needed. We have https://issues.redhat.com/browse/KEYCLOAK-16615 for that. I will approve once you have this migration test passing in the pipeline.

**Review by hmlnarik — APPROVED** [review]
LGTM, thanks @pedroigor ! The pipeline has revealed no regressions.
Please squash the two commits before merging.

**Comment by pedroigor** [comment]
@hmlnarik Going to send changes to documentation.

**Pull request merged**
